### PR TITLE
Fixed multi pipeline artifact download.

### PR DESCRIPTION
**Pull request by owenhuynMSFT**

### CODE REVIEW

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 1]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
...
+                options = DownloadPipelineArtifactOptions.CreateWithMultiManifestIds(
+                    artifactNameAndManifestId,
+                    targetDirectory,
+                    proxyUri: null,
+                    minimatchPatterns: minimatchFilters);
+            }
+            else
+            {
+                var buildArtifact = buildArtifacts.Single();
                 if (buildArtifact.Resource.Type != PipelineArtifactTypeName)
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: >buildArtifact [](start = 20, length = 13)

Needs a null check ?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 2]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
             string[] minimatchFilters,
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: >string[] [](start = 12, length = 8)

Should be IEnumerable,

because string[] contact implies you can modify the array elements

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 3]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
...
+            DownloadPipelineArtifactOptions options;
+            if (buildArtifacts.Count() > 1)
             {
+                IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
+                foreach (var buildArtifact in buildArtifacts)
+                {
+                    if (buildArtifact.Resource.Type != PipelineArtifactTypeName)
+                    {
+                        throw new ArgumentException("The artifact is not of the type Pipeline Artifact.");
+                    }
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: Move this validation outside of both if/else 
like :
if (buildArtifacts.Any(ba => ba.Resource.Type != PipelineArtifactTypeName)) { throw ... }

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 4]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
             string[] minimatchFilters,
             CancellationToken cancellationToken)
         {
-            IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
-            foreach (var buildArtifact in buildArtifacts)
+            DownloadPipelineArtifactOptions options;
+            if (buildArtifacts.Count() > 1)
             {
+                IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
+                foreach (var buildArtifact in buildArtifacts)
```
Comment (geekstew) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: Why not just do a where clause?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 5]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
...
+                // 2) download to the target path
+                options = DownloadPipelineArtifactOptions.CreateWithMultiManifestIds(
+                    artifactNameAndManifestId,
+                    targetDirectory,
+                    proxyUri: null,
+                    minimatchPatterns: minimatchFilters);
+            }
+            else
+            {
+                var buildArtifact = buildArtifacts.Single();
```
Comment (geekstew) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: SIngleOrDefault

Comment (geekstew) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: test for null

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 6]
```diff
@@ -171,22 +171,40 @@ private BuildDropManager CreateBulidDropManager(AgentTaskPluginExecutionContext
             string[] minimatchFilters,
             CancellationToken cancellationToken)
         {
-            IDictionary<string, DedupIdentifier> artifactNameAndManifestId = new Dictionary<string, DedupIdentifier>();
-            foreach (var buildArtifact in buildArtifacts)
+            DownloadPipelineArtifactOptions options;
+            if (buildArtifacts.Count() > 1)
```
Comment (geekstew) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: What if there are none?

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 5 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: We don't need to test for null, the client will throw if the buildArtifact is not found/doesn't exist

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 1 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: buildArtifact is never null because the client will fail the REST call if buildArtifact does not exist

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 6 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: Handled with 0 artifact check now

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 4 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: Changed to ToDictionary

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs (hunk 2 above)
Comment (owenhuynMSFT) [reply src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: Doesn't apply anymore

**Comment by arunkm** [comment]
>             if (string.IsNullOrEmpty(downloadParameters.ArtifactName))

We should get the multi-vs-single as a separate options/parameter rather than depending on name being empty
I am fine if this kind of check happens in Task/Presentation layer to pass that option (to keep UI same)

---
Refers to: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs:84 in 4789377. [](commit_id = 478937738a46b067057c641e1990d64d40b1789d, deletion_comment = False)

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs [hunk 7]
```diff
@@ -112,15 +113,20 @@ public class PipelineArtifactServer
...
                     context.Output(StringUtil.Loc("DownloadingMultiplePipelineArtifacts", pipelineArtifacts.Count()));
-                    await DownloadPipelineArtifacts(
-                        buildDropManager,
-                        pipelineArtifacts,
+
+                    var artifactNameAndManifestIds = pipelineArtifacts.ToDictionary(
+                        keySelector: (a) => a.Name, // keys should be unique, if not something is really wrong
+                        elementSelector: (a) => DedupIdentifier.Create(a.Resource.Data));
+                    // 2) download to the target path
+                    DownloadPipelineArtifactOptions options = DownloadPipelineArtifactOptions.CreateWithMultiManifestIds(
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs]: >DownloadPipelineArtifactOptions [](start = 20, length = 31)

use var, as it is obvious

File: src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs [hunk 8]
```diff
@@ -169,8 +169,18 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
 
             string fullPath = this.CreateDirectoryIfDoesntExist(downloadPath);
 
+            DownloadOptions downloadOptions;
```
Comment (arunkm) [inline src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs]: >downloadOptions [](start = 28, length = 15)

nit: you can use ternary operator here:

**Comment by Redno-Pnomis** [comment]
Is there an ETA on this fix making it into pre-release?

**Comment by TingluoHuang** [comment]
this is already part of 2.149.2 agent.
